packages/devtools_app/lib/src/framework/theme_manager.dart

```diff
@@ -0,0 +1,103 @@
+import 'dart:ui';
+
+import 'package:devtools_app_shared/ui.dart';
+import 'package:devtools_app_shared/utils.dart';
+import 'package:dtd/dtd.dart';
+import 'package:logging/logging.dart';
+
+import '../../devtools_app.dart';
+import '../service/editor/api_classes.dart';
+
+final _log = Logger('theme_manager');
+
+class ThemeManager {
+  ThemeManager(DartToolingDaemon dtd) : editorClient = DtdEditorClient(dtd);
+
+  final DtdEditorClient editorClient;
+
+  void listenForThemeChanges() {
+    editorClient.event.listen((event) {
+      if (event is ThemeChangedEvent) {
+        final currentTheme = getIdeTheme();
+        final newTheme = event.theme;
+
+        if (currentTheme.isDarkMode != newTheme.isDarkMode) {
+          currentTheme.isDarkMode = newTheme.isDarkMode;
+          updateQueryParameter(
+            IdeThemeQueryParams.devToolsThemeKey,
+            currentTheme.isDarkMode
+                ? IdeThemeQueryParams.darkThemeValue
+                : IdeThemeQueryParams.lightThemeValue,
+          );
+        }
+
+        if (newTheme.backgroundColor != null) {
+          final newBackgroundColor = _tryParseColor(newTheme.backgroundColor!);
+          if (newBackgroundColor != null &&
+              newBackgroundColor != currentTheme.backgroundColor) {
+            currentTheme.backgroundColor = newBackgroundColor;
+            updateQueryParameter(
+              IdeThemeQueryParams.backgroundColorKey,
+              _colorAsHex(newBackgroundColor),
+            );
+          }
+        }
+
+        if (newTheme.foregroundColor != null) {
+          final newForegroundColor = _tryParseColor(newTheme.foregroundColor!);
+          if (newForegroundColor != null &&
+              newForegroundColor != currentTheme.foregroundColor) {
+            currentTheme.foregroundColor = newForegroundColor;
+            updateQueryParameter(
+              IdeThemeQueryParams.foregroundColorKey,
+              _colorAsHex(newForegroundColor),
+            );
+          }
+        }
+
+        if (newTheme.fontSize != null &&
+            newTheme.fontSize!.toDouble() != currentTheme.fontSize) {
+          currentTheme.fontSize = newTheme.fontSize!.toDouble();
```

Comment:
nit: instead of tracking a new theme variable, can we just update the query parameters and then call `setGlobal(IdeTheme, getIdeTheme())` again to set the theme with the updated query params?